### ISSUE TITLE
LocaleUtil::formatAsLocale() must be of the type string, null returned

### ISSUE THREAD

**Issue by bizon1** [contao/contao#3593]
**Affected version(s)**

4.12.3

**Description**

Since I updated a local contao testproject from version 4.12.2 to 4.12.3 an error message appears on each site which has an active Isotope listing module(mod_iso_productlist and iso_list_default.html5).

**var/logs:**

`[2021-10-20T23:31:50.589962+02:00] request.INFO: Matched route "tl_page.30". {"route":"tl_page.30","route_parameters":{"_token_check":true,"_controller":"Contao\\FrontendIndex::renderPage","_scope":"frontend","_locale":"de","_format":"html","_canonical_route":"tl_page.30","parameters":"","pageModel":{"Contao\\PageModel":[]},"_route":"tl_page.30","_route_object":{"Contao\\CoreBundle\\Routing\\Page\\PageRoute":[]}},"request_uri":"https://example.local/shop/parts/","method":"GET"} []

[2021-10-20T23:31:50.614605+02:00] security.INFO: Populated the TokenStorage with an anonymous Token. [] []

[2021-10-20T23:31:50.955490+02:00] request.CRITICAL: Uncaught PHP Exception TypeError: "Return value of Contao\CoreBundle\Util\LocaleUtil::formatAsLocale() must be of the type string, null returned" at /path/to/project/vendor/contao/core-bundle/src/Util/LocaleUtil.php line 112 {"exception":"[object] (TypeError(code: 0): Return value of Contao\\CoreBundle\\Util\\LocaleUtil::formatAsLocale() must be of the type string, null returned at /path/to/project/vendor/contao/core-bundle/src/Util/LocaleUtil.php:112)"} []`

**Stack Trace:**
`TypeError:
Return value of Contao\CoreBundle\Util\LocaleUtil::formatAsLocale() must be of the type string, null returned
  at vendor/contao/core-bundle/src/Util/LocaleUtil.php:112

  at Contao\CoreBundle\Util\LocaleUtil::formatAsLocale()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:1233)
  at Contao\Controller::generateFrontendUrl()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/Product/Standard.php:1084)
  at Isotope\Model\Product\Standard->generateUrl()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/Product/Standard.php:614)
  at Isotope\Model\Product\Standard->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/ProductList.php:264)
  at Isotope\Module\ProductList->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/codefog/contao-haste/library/Haste/Frontend/AbstractFrontendModule.php:52)
  at Haste\Frontend\AbstractFrontendModule->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/Module.php:112)
  at Isotope\Module\Module->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/ProductList.php:96)
  at Isotope\Module\ProductList->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:101)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:595)
  at Contao\Controller::getContentElement()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:202)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:74)
  at Contao\ModuleArticle->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:523)
  at Contao\Controller::getArticle()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:365)
  at Contao\Controller::getFrontendModule()
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:184)
  at Contao\PageRegular->prepare()
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:58)
  at Contao\PageRegular->getResponse()
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:339)
  at Contao\FrontendIndex->renderPage()
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (web/index.php:31)
`

Downgrade to contao-4.12.2 is fixing the issue for now.

**Comment by aschempp** [comment]
Isotope seems to pass the page language to `Controller::generateFrontendUrl()`, not sure why that is empty. @bizon1 not sure if you're a developer, can you debug what `$objJumpTo->language` is in `system/modules/isotope/library/Isotope/Model/Product/Standard.php` around line 1082?

**Comment by fritzmg** [comment]
May be a `->loadDetails()` is missing now in Contao 4.12.3 for the target page?

**Comment by bizon1** [comment]
Hi Andreas,
`var_dump($objJumpTo->language)` delivers `string(0) ""`

**Comment by fritzmg** [comment]
What if you add `$objJumpTo->loadDetails();` in line 1081?

**Comment by bizon1** [comment]
> What if you add $objJumpTo->loadDetails(); in line 1081?

Thx Fritz,
this fixed the issue.
`var_dump($objJumpTo->language)` is giving now `string(2) "de"`

**Comment by leofeyer** [comment]
So this is an Isotope issue?

**Comment by bizon1** [comment]
It seems to be.
Just adding `$objJumpTo->loadDetails();` at line 1081 in `vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/Product/Standard.php` solved it.

**Comment by aschempp** [comment]
it is both an Isotope and a core issue to me, I'll do a PR 😊

**Comment by fritzmg** [comment]
This particular error is likely fixed by #3682, however Isotope probably still should add the `loadDetails` call.

**Comment by markocupic** [comment]
Similar error under Isotope 2.8.7 & Contao 4.13.12 when running the Isotope Orders Module https://xxx.de/contao?do=iso_orders&act=edit&id=xxx in the backend. Any idea? I'm not sure, if it is a Contao or Isotope issue.

```
TypeError:
Argument 1 passed to Contao\CoreBundle\Util\LocaleUtil::formatAsLocale() must be of the type string, null given, called in /kunden/633798_50670/webseiten/schulfilme_backup.com/vendor/contao/core-bundle/src/Resources/contao/library/Contao/InsertTags.php on line 1530

  at vendor/contao/core-bundle/src/Util/LocaleUtil.php:109
  at Contao\CoreBundle\Util\LocaleUtil::formatAsLocale()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/InsertTags.php:1530)
  at Contao\InsertTags->languageMatches()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/InsertTags.php:761)
  at Contao\InsertTags->executeReplace()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/InsertTags.php:100)
  at Contao\InsertTags->replaceInternal()
     (vendor/contao/core-bundle/src/InsertTag/InsertTagParser.php:73)
  at Contao\CoreBundle\InsertTag\InsertTagParser->callLegacyClass()
     (vendor/contao/core-bundle/src/InsertTag/InsertTagParser.php:37)
  at Contao\CoreBundle\InsertTag\InsertTagParser->replaceChunked()
     (vendor/contao/core-bundle/src/InsertTag/InsertTagParser.php:32)
  at Contao\CoreBundle\InsertTag\InsertTagParser->replace()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:842)
  at Contao\Controller::replaceInsertTags()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Isotope.php:484)
  at Isotope\Isotope::formatOptions()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/ProductCollection.php:1682)
  at Isotope\Model\ProductCollection->generateItem()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/ProductCollection/Order.php:570)
  at Isotope\Model\ProductCollection\Order->addItemsToTemplate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Model/ProductCollection.php:1492)
  at Isotope\Model\ProductCollection->addToTemplate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/AbstractProductCollection.php:90)
  at Isotope\Module\AbstractProductCollection->compile()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/OrderDetails.php:69)
  at Isotope\Module\OrderDetails->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/codefog/contao-haste/library/Haste/Frontend/AbstractFrontendModule.php:52)
  at Haste\Frontend\AbstractFrontendModule->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/Module.php:115)
  at Isotope\Module\Module->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/AbstractProductCollection.php:59)
  at Isotope\Module\AbstractProductCollection->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Module/OrderDetails.php:55)
  at Isotope\Module\OrderDetails->generate()
     (vendor/isotope/isotope-core/system/modules/isotope/library/Isotope/Backend/ProductCollection/Callback.php:125)
  at Isotope\Backend\ProductCollection\Callback->generateOrderDetails()
     (vendor/contao/core-bundle/src/Resources/contao/classes/DataContainer.php:384)
  at Contao\DataContainer->row()
     (vendor/contao/core-bundle/src/Resources/contao/drivers/DC_Table.php:1973)
  at Contao\DC_Table->edit()
     (vendor/contao/core-bundle/src/Resources/contao/classes/Backend.php:667)
  at Contao\Backend->getBackendModule()
     (vendor/contao/core-bundle/src/Resources/contao/controllers/BackendMain.php:168)
  at Contao\BackendMain->run()
     (vendor/contao/core-bundle/src/Controller/BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor/symfony/http-kernel/HttpKernel.php:153)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (web/index.php:44)                
```
